File: packages/typescript/scripts/release-notes.mjs
import { readFile, writeFile } from "node:fs/promises";
import { pathToFileURL } from "node:url";

export const extractReleaseNotes = (changelog, version) => {
  if (typeof version !== "string" || version.trim().length === 0) {
    throw new Error("A non-empty release version is required.");
  }

  const headings = [...changelog.matchAll(/^##[ \t]+(.+?)[ \t]*$/gm)];
  const matches = headings.filter((heading) => heading[1] === version);

  if (matches.length === 0) {
    throw new Error(`CHANGELOG.md does not contain a \"## ${version}\" section.`);
  }

  if (matches.length > 1) {
    throw new Error(`CHANGELOG.md contains more than one \"## ${version}\" section.`);
  }

  const heading = matches[0];
  const headingIndex = headings.indexOf(heading);
  const nextHeading = headings[headingIndex + 1];
  const start = (heading.index ?? 0) + heading[0].length;
  const end = nextHeading?.index ?? changelog.length;
  const notes = changelog.slice(start, end).trim();

  if (notes.length === 0) {
    throw new Error(`The \"## ${version}\" section in CHANGELOG.md is empty.`);
  }

  return notes;
};

const run = async () => {
  const [, , version, outputPath] = process.argv;

  if (!version || !outputPath) {
    throw new Error("Usage: node scripts/release-notes.mjs <version> <output-path>");
  }

  const changelog = await readFile(new URL("../CHANGELOG.md", import.meta.url), "utf8");
  const notes = extractReleaseNotes(changelog, version);
  await writeFile(outputPath, `${notes}\n`);
};

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  await run();
}

File: packages/python/scripts/release_notes.py
from __future__ import annotations

import re
import sys
from pathlib import Path


def extract(version: str, changelog: str) -> str:
    matches = list(re.finditer(rf"^## {re.escape(version)}\s*$", changelog, re.MULTILINE))
    if len(matches) != 1:
        qualifier = "missing" if not matches else "duplicate"
        raise ValueError(f"Changelog section for {version} is {qualifier}.")
    start = matches[0].end()
    next_heading = re.search(r"^## \S+\s*$", changelog[start:], re.MULTILINE)
    end = start + next_heading.start() if next_heading else len(changelog)
    notes = changelog[start:end].strip()
    if not notes:
        raise ValueError(f"Changelog section for {version} is empty.")
    return notes + "\n"


def main() -> None:
    if len(sys.argv) != 3:
        raise SystemExit("usage: release_notes.py <version> <output-path>")
    changelog = (Path(__file__).parents[1] / "CHANGELOG.md").read_text()
    Path(sys.argv[2]).write_text(extract(sys.argv[1], changelog))


if __name__ == "__main__":
    main()
